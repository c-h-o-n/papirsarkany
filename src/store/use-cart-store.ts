import 'client-only';
import { create } from 'zustand';
import { createJSONStorage, persist } from 'zustand/middleware';

import type { ShippingFee } from '~/lib/types';
import type { CartItem } from '~/lib/validation-schemas';

type State = {
  _hasHydrated: boolean;
  cart: CartItem[];
  shippingFee: ShippingFee;
  billingFee: number;
};

type Actions = {
  setHasHydrated: (hasHydrated: boolean) => void;
  addToCart: (item: CartItem) => void;
  removeFromCart: (item: CartItem) => void;
  increaseItemQuantity: (item: CartItem) => void;
  decreaseItemQuantity: (item: CartItem) => void;
  setItemQuantity: (item: CartItem, quantity: number) => void;
  setShippingFee: (shippingFee: ShippingFee) => void;
  setBillingFee: (billingFee: number) => void;
  resetCart: () => void;
};

const initialState: State = {
  _hasHydrated: false,
  cart: [],
  shippingFee: 0,
  billingFee: 0,
};

export const useCartStore = create(
  persist<State & Actions>(
    (set, get) => ({
      ...initialState,
      setHasHydrated: (state) => {
        set({
          _hasHydrated: state,
        });
      },
      addToCart(productToAdd) {
        const { cart, increaseItemQuantity } = get();
        const cartItem = cart.find(
          (item) =>
            item._id === productToAdd._id && item.name === productToAdd.name,
        );

        if (cartItem) {
          increaseItemQuantity(cartItem);
          return;
        }

        const newProductToAdd: CartItem = { ...productToAdd, quantity: 1 };

        set(() => ({
          cart: [...cart, newProductToAdd],
        }));
      },
      increaseItemQuantity(item) {
        item.quantity++;

        set((state) => ({ cart: [...state.cart] }));
      },
      decreaseItemQuantity(product) {
        set((state) => ({
          cart: state.cart
            .map((item) =>
              item._id === product._id
                ? { ...item, quantity: item.quantity - 1 }
                : item
            )
            .filter((item) => item.quantity > 0), // Remove item if quantity is 0
        }))
      },
      setItemQuantity(product, quantity) {
        if (quantity < 1 || quantity > 999) {
          return;
        }

        product.quantity = quantity;

        set((state) => ({
          cart: [...state.cart],
        }));
      },
      removeFromCart(product) {
        set((state) => ({
          cart: state.cart.filter(
            (item) => item._id !== product._id || item.name !== product.name,
          ),
        }));
      },
      setShippingFee(shippingFee) {
        set({
          shippingFee,
        });
      },
      setBillingFee(billingFee) {
        set({
          billingFee,
        });
      },
      resetCart() {
        set((state) => ({
          ...initialState,
          _hasHydrated: state._hasHydrated,
        }));
      },
    }),
    {
      onRehydrateStorage: () => (state) => {
        state?.setHasHydrated(true);
      },
      name: 'cart-storage',
      storage: createJSONStorage(() => sessionStorage),
      skipHydration: true,
    },
  ),
);
