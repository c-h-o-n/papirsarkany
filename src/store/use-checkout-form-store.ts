import 'client-only';
import { create } from 'zustand';

import type { OrderForm } from '~/lib/validation-schemas';

type State = {
  /**
   * Stored values of checkout form.
   */
  formValues: Partial<OrderForm>;
  step: number;
};

type Actions = {
  setFormValues: (formData: Partial<OrderForm>) => void;
  nextStep: () => void;
  prevStep: () => void;
  setStep: (step: number) => void;
  resetForm: () => void;
};

const initialState: State = {
  step: 0,
  formValues: {
    isSameAdressAsShipping: true,
  },
};

/**
 * To persist form state when user navigates to other pages during checkout and the checkout form got unmounted.
 */
export const useCheckoutFormStore = create<State & Actions>((set) => ({
  ...initialState,
  setFormValues(formData) {
    set((state) => ({ formValues: { ...state.formValues, ...formData } }));
  },
  nextStep() {
    set((state) => ({ step: state.step + 1 }));
  },
  prevStep() {
    set((state) => ({ step: state.step - 1 }));
  },
  setStep(step) {
    set(() => ({ step }));
  },
  resetForm() {
    set(() => ({
      ...initialState,
    }));
  },
}));
