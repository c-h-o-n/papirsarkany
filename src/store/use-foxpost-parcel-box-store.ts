import "client-only";
import { create } from "zustand";

type State = {
  destination?: string;
};
type Actions = {
  setFoxpostData: (data: State) => void;
};

const initialState: State = {
  destination: undefined,
};

export const useFoxpostParcelBoxStore = create<State & Actions>((set) => ({
  ...initialState,
  setFoxpostData(data) {
    set((state) => ({ ...state, ...data }));
  },
}));
