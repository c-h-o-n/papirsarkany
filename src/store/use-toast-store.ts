import "client-only";
import type { SetOptional } from "type-fest";
import { create } from "zustand";
import { delay } from "~/lib/helpers";
import type { Toast } from "~/lib/types";

/**
 * number of ms to wait before automatically closing the toast
 */
const AUTO_CLOSE = 3000;
/**
 * number of ms of framer motion animation duration
 */
const ANIMATION_DURATION = 300;

type State = {
  toasts: Toast[];
};

type Actions = {
  /**
   * Add a new toast to the store
   */
  toast: (toast: SetOptional<Toast, "id">) => void;
  /**
   * Remove a toast from the store
   */
  removeToast: (id: string) => void;
  /**
   * Set the next toast to active
   */
  showNextToast: () => void;

  /**
   * Set the toast to inactive
   */
  dismissToast: (id: string) => void;
};

const initialState: State = {
  toasts: [],
};

export const useToastStore = create<State & Actions>()((set, get) => ({
  ...initialState,

  toast: (toast) => {
    const { toasts, showNextToast } = get();

    const toastId = toast.id || crypto.randomUUID();

    if (toasts.find((toast) => toast.id === toastId)) {
      return;
    }

    set((state) => ({
      toasts: [...state.toasts, { ...toast, id: toastId, active: false }],
    }));

    if (!toasts.some((toast) => toast.active)) {
      showNextToast();
    }
  },

  removeToast: (id) => {
    set((state) => ({
      toasts: state.toasts.filter((toast) => toast.id !== id),
    }));
  },

  showNextToast: async () => {
    const { toasts, removeToast, dismissToast, showNextToast } = get();

    if (toasts.length === 0) {
      return;
    }

    const currentToast = toasts[0];

    set((state) => ({
      toasts: state.toasts.map((toast) =>
        toast.id === currentToast.id ? { ...toast, active: true } : toast,
      ),
    }));

    await delay(AUTO_CLOSE);

    dismissToast(currentToast.id);

    await delay(ANIMATION_DURATION);

    removeToast(currentToast.id);

    if (toasts.length > 0) {
      showNextToast();
    }
  },

  dismissToast: async (id) => {
    set((state) => ({
      toasts: state.toasts.map((toast) =>
        toast.id === id ? { ...toast, active: false } : toast,
      ),
    }));
  },
}));
