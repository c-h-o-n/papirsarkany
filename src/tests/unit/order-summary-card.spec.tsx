import { render, renderHook, screen } from "@testing-library/react";
import { expect, test } from "vitest";
import OrderSummaryCard from "~/components/order-summary-card";
import { kiteMock } from "~/mocks/product.mock";
import { useCartStore } from "~/store/use-cart-store";

test('should render "Üres a kosarad" when cart is empty', () => {
  renderHook(() => useCartStore.setState({ cart: [] }));

  const { container } = render(<OrderSummaryCard />);

  expect(screen.getByText("Üres a kosarad.")).toBeDefined();
  expect(container).toMatchSnapshot();
});

test('should render "definitive" layout when provided', () => {
  renderHook(() =>
    useCartStore.setState({
      cart: [
        {
          ...kiteMock,
          quantity: 2,
        },
      ],
    }),
  );

  const { container } = render(<OrderSummaryCard layout="definitive" />);
  expect(screen.queryByTestId("cart-item-controls")).toBeNull();
  expect(container).toMatchSnapshot();
});

test('should render "full" layout by default', () => {
  renderHook(() =>
    useCartStore.setState({
      cart: [
        {
          ...kiteMock,
          quantity: 2,
        },
      ],
    }),
  );

  const { container } = render(<OrderSummaryCard />);
  expect(screen.queryByTestId("cart-item-controls")).toBeDefined();
  expect(container).toMatchSnapshot();
});
