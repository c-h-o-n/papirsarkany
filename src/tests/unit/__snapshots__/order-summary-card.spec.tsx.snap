// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render "definitive" layout when provided 1`] = `
<div>
  <div
    class="rounded-lg border-4 border-black bg-white flex flex-1 flex-col gap-4 p-8"
  >
    <div
      class="flex flex-wrap items-center justify-between gap-2"
    >
      <div
        class="flex flex-shrink gap-2"
      >
        <div
          class="hidden flex-shrink-0 min-[390px]:block"
        >
          <img
            alt="Ördögfióka"
            class="h-auto max-h-32 min-h-24 w-32 rounded-lg object-contain"
            data-nimg="1"
            decoding="async"
            height="128"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fcdn.sanity.io%2Fimages%2Fs2hk1b39%2Fdevelop%2Fa659f9fef018cf839676e3c961eaea12d57205b4-800x600.webp&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fcdn.sanity.io%2Fimages%2Fs2hk1b39%2Fdevelop%2Fa659f9fef018cf839676e3c961eaea12d57205b4-800x600.webp&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fcdn.sanity.io%2Fimages%2Fs2hk1b39%2Fdevelop%2Fa659f9fef018cf839676e3c961eaea12d57205b4-800x600.webp&w=256&q=75 2x"
            style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
            width="128"
          />
        </div>
        <div>
          <h3
            class="font-bold"
          >
            Ördögfióka
          </h3>
          <span
            class="text-sm font-normal text-gray-400"
          >
            2
             db
          </span>
        </div>
      </div>
      <div
        class="flex flex-1 items-center justify-end gap-4"
      >
        <h3
          class="font-bold"
        >
          3 000 Ft
        </h3>
      </div>
    </div>
    <div
      class="d-divider"
    />
    <div
      class="flex justify-between font-bold"
    >
      <h3
        class="text-balance"
      >
        Összesen
         
        <span
          class="text-base font-normal text-gray-400"
        >
          2
           db
        </span>
      </h3>
      <h3>
        3 000 Ft
      </h3>
    </div>
  </div>
</div>
`;

exports[`should render "full" layout by default 1`] = `
<div>
  <div
    class="rounded-lg border-4 border-black bg-white flex flex-1 flex-col gap-4 p-8"
  >
    <div
      class="flex items-center justify-between gap-2"
    >
      <div
        class="flex gap-2"
      >
        <img
          alt="Ördögfióka"
          class="h-auto max-h-32 min-h-24 w-32 rounded-lg object-contain"
          data-nimg="1"
          decoding="async"
          height="128"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fcdn.sanity.io%2Fimages%2Fs2hk1b39%2Fdevelop%2Fa659f9fef018cf839676e3c961eaea12d57205b4-800x600.webp&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fcdn.sanity.io%2Fimages%2Fs2hk1b39%2Fdevelop%2Fa659f9fef018cf839676e3c961eaea12d57205b4-800x600.webp&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fcdn.sanity.io%2Fimages%2Fs2hk1b39%2Fdevelop%2Fa659f9fef018cf839676e3c961eaea12d57205b4-800x600.webp&w=256&q=75 2x"
          style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
          width="128"
        />
        <div>
          <h3
            class="font-bold"
          >
            Ördögfióka
          </h3>
          <h4
            class="block font-bold md:hidden"
          >
            3 000 Ft
          </h4>
        </div>
      </div>
      <div
        class="hidden flex-1 items-center justify-end gap-4 md:flex"
      >
        <h4
          class="font-bold"
        >
          3 000 Ft
        </h4>
        <div>
          <div
            class="flex items-center"
          >
            <div
              class="d-btn no-animation rounded-r-none shadow-none"
            >
              -
            </div>
            <input
              class="h-12 w-12 rounded-none bg-base-200 text-center shadow-none"
              type="number"
              value="2"
            />
            <div
              class="d-btn no-animation rounded-l-none shadow-none false"
            >
              +
            </div>
          </div>
        </div>
        <button
          class="d-btn d-btn-square d-btn-error"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            height="800px"
            viewBox="0 0 24 24"
            width="800px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 12V17"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M14 12V17"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M4 7H20"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M6 10V18C6 19.6569 7.34315 21 9 21H15C16.6569 21 18 19.6569 18 18V10"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M9 5C9 3.89543 9.89543 3 11 3H13C14.1046 3 15 3.89543 15 5V7H9V5Z"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex justify-between gap-4 md:hidden"
      data-testid="cart-item-controls"
    >
      <div
        class="flex items-center"
      >
        <div
          class="d-btn no-animation rounded-r-none shadow-none"
        >
          -
        </div>
        <input
          class="h-12 w-12 rounded-none bg-base-200 text-center shadow-none"
          type="number"
          value="2"
        />
        <div
          class="d-btn no-animation rounded-l-none shadow-none false"
        >
          +
        </div>
      </div>
      <button
        class="d-btn d-btn-square d-btn-error"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          height="800px"
          viewBox="0 0 24 24"
          width="800px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 12V17"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M14 12V17"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M4 7H20"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M6 10V18C6 19.6569 7.34315 21 9 21H15C16.6569 21 18 19.6569 18 18V10"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M9 5C9 3.89543 9.89543 3 11 3H13C14.1046 3 15 3.89543 15 5V7H9V5Z"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
    <div
      class="d-divider"
    />
    <div
      class="flex justify-between font-bold"
    >
      <h3>
        Összesen
         
        <span
          class="text-base font-normal text-gray-400"
        >
          2
           db
        </span>
      </h3>
      <h3>
        3 000 Ft
      </h3>
    </div>
  </div>
</div>
`;

exports[`should render "Üres a kosarad" when cart is empty 1`] = `
<div>
  <div
    class="text-center"
  >
    <h1>
      Üres a kosarad.
    </h1>
  </div>
</div>
`;
